chain/consensus/mir/state_manager.go

```diff
@@ -0,0 +1,215 @@
+package mir
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/multiformats/go-multiaddr"
+
+	"github.com/filecoin-project/mir/pkg/pb/checkpointpb"
+	"github.com/filecoin-project/mir/pkg/pb/commonpb"
+	"github.com/filecoin-project/mir/pkg/pb/requestpb"
+	"github.com/filecoin-project/mir/pkg/systems/smr"
+	t "github.com/filecoin-project/mir/pkg/types"
+	"github.com/filecoin-project/mir/pkg/util/maputil"
+)
+
+const (
+	availabilityModuleID = t.ModuleID("availability")
+)
+
+var _ smr.AppLogic = &StateManager{}
+
+type Message []byte
+
+type Batch struct {
+	Messages   []Message
+	Validators []t.NodeID
+}
+
+type StateManager struct {
+
+	// The current epoch number.
+	currentEpoch t.EpochNr
+
+	// For each epoch number, stores the corresponding membership.
+	memberships map[t.EpochNr]map[t.NodeID]t.NodeAddress
+
+	// Channel to send batches to Lotus.
+	NextBatch chan *Batch
+
+	// Channel to send a membership.
+	NewMembership chan map[t.NodeID]t.NodeAddress
+
+	MirManager *Manager
+
+	reconfigurationVotes map[t.EpochNr]map[string]int
+}
+
+func NewStateManager(initialMembership map[t.NodeID]t.NodeAddress, m *Manager) *StateManager {
+	// Initialize the membership for the first epochs.
+	// We use configOffset+2 memberships to account for:
+	// - The first epoch (epoch 0)
+	// - The configOffset epochs that already have a fixed membership (epochs 1 to configOffset)
+	// - The membership of the following epoch (configOffset+1) initialized with the same membership,
+	//   but potentially replaced during the first epoch (epoch 0) through special configuration requests.
+	memberships := make(map[t.EpochNr]map[t.NodeID]t.NodeAddress, ConfigOffset+2)
+	for e := 0; e < ConfigOffset+2; e++ {
+		memberships[t.EpochNr(e)] = initialMembership
+	}
+
+	sm := StateManager{
+		NextBatch:            make(chan *Batch),
+		NewMembership:        make(chan map[t.NodeID]t.NodeAddress, 1),
+		MirManager:           m,
+		memberships:          memberships,
+		currentEpoch:         0,
+		reconfigurationVotes: make(map[t.EpochNr]map[string]int),
+	}
+	return &sm
+}
+
+// RestoreState restores the application's state to the one represented by the passed argument.
+// The argument is a binary representation of the application state returned from Snapshot().
+func (sm *StateManager) RestoreState(snapshot []byte, config *commonpb.EpochConfig) error {
+	sm.currentEpoch = t.EpochNr(config.EpochNr)
+	sm.memberships = make(map[t.EpochNr]map[t.NodeID]t.NodeAddress, len(config.Memberships))
+
+	for e, membership := range config.Memberships {
+		sm.memberships[t.EpochNr(e)] = make(map[t.NodeID]t.NodeAddress)
+		for nID, nAddr := range membership.Membership {
+			var err error
+			sm.memberships[t.EpochNr(e)][t.NodeID(nID)], err = multiaddr.NewMultiaddr(nAddr)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
You can use the `t.Membership` function to convert from pb.

Reply:
Fixed